### PR TITLE
Adds Form ratio input and alignItems prop to Grid.Row

### PR DESCRIPTION
**Pull request by jonthomp**
A simple new prop for centring items of Grid.Row along with a new form ratio input component.

The ratio component can be controlled or uncontrolled through the use of defaultValue, value and onChange props.

screenshot from 2018-04-23 17-46-46

### CODE REVIEW

File: src/components/Form/FormRatio.react.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  +onChange?: (SyntheticInputEvent<EventTarget>) => mixed,
+|};
+
+type State = {|
+  internalValue: number,
+  hasChanged: boolean,
+|};
+
+class FormRatio extends React.PureComponent<Props, State> {
+  constructor(props: Props) {
```
Comment (AaronCoplan) [inline src/components/Form/FormRatio.react.js]: You can actually replace this whole `constructor(...)` block with the following:

```
state = {
    internalValue: this.props.defaultValue || this.props.value || 0,
    hasChanged: false,
};
```

File: src/components/Form/FormRatio.react.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  handleOnChange = (e: SyntheticInputEvent<EventTarget>) => {
+    const { onChange = false } = this.props;
+    const value: number = Number(e.target.value);
+    if (onChange) {
+      onChange(e);
+      this.setState({ hasChanged: true });
+    } else {
+      this.setState({ internalValue: value, hasChanged: true });
+    }
+  };
```
Comment (AaronCoplan) [inline src/components/Form/FormRatio.react.js]: [nit] insert empty line after this

File: src/components/Form/FormRatio.react.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  render() {
```
Comment (AaronCoplan) [inline src/components/Form/FormRatio.react.js]: Add return type `render(): React.Node {`

File: src/components/Form/FormRatio.react.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  +min?: number,
+  +max?: number,
+  +defaultValue?: number,
+  +value?: number,
+  +onChange?: (SyntheticInputEvent<EventTarget>) => mixed,
+|};
+
+type State = {|
+  internalValue: number,
+  hasChanged: boolean,
```
Comment (AaronCoplan) [inline src/components/Form/FormRatio.react.js]: Not totally following what `hasChanged` is used for.  Can you just kind of explain the logic with `onChange` and these state variables?

File: src/components/Form/FormRatio.react.js (hunk 2 above)
Comment (jonthomp) [reply src/components/Form/FormRatio.react.js]: This appears to be something prettier is doing, shall investigate an option to change it

**Comment by jonthomp** [comment]
@AaronCoplan I've tried to make the logic clearer to explain how it works, and in doing so removed hasChanged from state, but I know there is a better way of typing and doing this with disjoint unions. I haven't yet been able to get it all working together. How would you go about it?

**Comment by jonthomp** [comment]
@AaronCoplan I've made progress with the disjoint union typing, it now works, and in doing so have simplified the internals of the component class, other requested amends also complete, ready for review whenever is good for you :)

**Comment by AaronCoplan** [comment]
Awesome, this looks great!  Love the use of multiple types there - super easy to read and understand now 😀

**Pull request merged**